Consolidate accounts onboarding: Create the basic combo card for connecting Google account

ankitrox opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #2566  .

_Replace this with a good description of your changes & reasoning._


### Screenshots:


<img width="1143" alt="Screenshot 2024-09-11 at 5 53 28 PM" src="https://github.com/user-attachments/assets/f35163c7-33ba-41b8-901e-d3246d3cbdc3">

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Disconnect all Google accounts if already connected in the plugin.

2. Start the onboarding process again.

3. There should be Google card visible in the first step.

4. The connect button would be disabled if Jetpack is not connected or Terms and conditions are not accepted.

5. Clicking the "Connect" button initiates the current flow to connect a Google account and grant permissions

6. After connecting the Google account successfully, the connected version of Google card should be visible.


### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Update - Introduce new UI for Google accounts setup during the onboarding.

ankitrox committed:
Onboarding: Create New Google Combo Accounts Card.

ankitrox committed:
Fix linting issues.

ankitrox committed:
Fix linting issues.

ankitrox committed:
Fix js lint.

ankitrox committed:
Disable eslint error.

ankitrox committed:
Resolve conflicts with base.

ankitrox committed:
Fix failing tests.

Remove ads and merchant center tests.

joemcgill commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,119 @@
+			),
+			{
+				linkMerchant: (
+					<AppDocumentationLink
+						context="setup-mc-accounts"
+						linkId="required-google-permissions"
+						href="https://woocommerce.com/document/google-for-woocommerce/get-started/setup-and-configuration/#required-google-permissions"
+					/>
```
The previous modal for accepting MC terms looked like this:
```suggestion
					<AppDocumentationLink
						context="setup-mc"
						linkId="google-mc-terms-of-service"
						href="https://support.google.com/merchants/answer/160173"
					/>
```

joemcgill replied:
Handled in [85dadb0](https://github.com/woocommerce/google-listings-and-ads/pull/2601/commits/85dadb0cbff6aab7931a9d9353b2e42bfbab5747)

joemcgill commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,119 @@
+					/>
+				),
+				linkAds: (
+					<AppDocumentationLink
+						context="setup-mc-accounts"
+						linkId="required-google-permissions"
+						href="https://woocommerce.com/document/google-for-woocommerce/get-started/setup-and-configuration/#required-google-permissions"
+					/>
```
The previous link in the terms modal for creating an Ads account looked like this:
```suggestion
					<AppDocumentationLink
						context="setup-ads"
						linkId="google-ads-terms-of-service"
						href="https://support.google.com/adspolicy/answer/54818"
					/>
```

joemcgill replied:
Handled in [85dadb0](https://github.com/woocommerce/google-listings-and-ads/pull/2601/commits/85dadb0cbff6aab7931a9d9353b2e42bfbab5747)

joemcgill commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,119 @@
+									'google-listings-and-ads'
+								),
+								{
+									link: readMoreLink,
```
👍🏻 exactly

joemcgill reviewed (commented):
This is starting to look really good. I've left some feedback inline about the Terms & Conditions links that were previously used.

ankitrox committed:
Update links.

joemcgill commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,119 @@
+ * @param {boolean} props.disabled
+ * @fires gla_google_account_connect_button_click with `{ action: 'authorization', context: 'reconnect' }`
+ * @fires gla_google_account_connect_button_click with `{ action: 'authorization', context: 'setup-mc' }`
+ * @fires gla_documentation_link_click with `{ context: 'setup-mc-accounts', link_id: 'required-google-permissions', href: 'https://woocommerce.com/document/google-for-woocommerce/get-started/setup-and-configuration/#required-google-permissions' }`
```
This accounts for the `readMoreLink` component. Will need to add two additional ones for the two new `AppDocumentationLink` components.

joemcgill commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,119 @@
+						/>
+						{
+							// eslint-disable-next-line jsx-a11y/label-has-associated-control
+							<label htmlFor={ checkboxId }>
+								<CheckboxLabel />
+							</label>
```
Can the `CheckboxLabel` component be directly passed to the `label` prop of the `CheckboxControl`?

ankitrox replied:
It won't be possible to move directly to `label` prop because it expects a string, but `createInterpolateElement` would return react element, thus it will cause the error.

asvinb replied:
@ankitrox I tried it locally and it seems to be working fine.
```js
<CheckboxControl
	checked={ termsAccepted }
	onChange={ ( val ) => setTermsAccepted( !! val ) }
	disabled={ disabled }
	label={ createInterpolateElement(
		__(
			'I accept the terms and conditions of <linkMerchant>Merchant Center</linkMerchant> and <linkAds>Google Ads</linkAds>',
			'google-listings-and-ads'
		),
		{
			linkMerchant: (
				<AppDocumentationLink
					context="setup-mc-accounts"
					linkId="google-mc-terms-of-service"
					href="https://support.google.com/merchants/answer/160173"
				/>
			),
			linkAds: (
				<AppDocumentationLink
					context="setup-ads"
					linkId="google-ads-terms-of-service"
					href="https://support.google.com/adspolicy/answer/54818"
				/>
			),
		}
	) }
/>
```

Screenshot:
![image](https://github.com/user-attachments/assets/0b857c72-64e4-44d2-ae58-387019609bba)

Can you kindly double check on your end please? Thanks!

ankitrox replied:
@asvinb Wow!

This seems to be working. It was throwing error for me yesterday, can't really remember what was wrong. Moreover; looking at the [component doc](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/checkbox-control), it seems that it only supports string, but this seems to be working, so thanks for checking this :)

joemcgill reviewed (commented):
Left a couple more notes, but this is looking right to me besides updating the E2E tests. I think @ankitguptaindia could start some QA testing while the E2E updates are being made.

ankitrox committed:
Add E2E tests for combo card.

ankitrox committed:
Styling changes

asvinb commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,116 @@
+	return (
+		<AccountCard
+			appearance={ APPEARANCE.GOOGLE }
+			className={ 'google-combo-account-card' }
```
```suggestion
			className="google-combo-account-card"
```

asvinb commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,116 @@
+						/>
+						{
+							// eslint-disable-next-line jsx-a11y/label-has-associated-control
+							<label htmlFor={ checkboxId }>
```
@ankitrox No need to have a separate label element for the checkbox here. We can use the `label` prop of the `CheckboxControl` component.

ankitrox replied:
Please check [this comment](https://github.com/woocommerce/google-listings-and-ads/pull/2601#discussion_r1756770687).

asvinb commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,116 @@
+			alignIcon="top"
+			description={
+				<>
+					{ __(
```
@ankitrox Let's wrap this string within a paragraph tag.

ankitrox replied:
@asvinb If we wrap it in paragraph tag, it will produce invalid HTML because checkbox control will yield `div` and we should not have div inside `<p>` tag.

asvinb replied:
Apologies. I was unclear @ankitrox . Only the text `Required to sync with Google Merchant Center and Google Ads.` should be wrapped within a `p` tag.

asvinb commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,116 @@
+						'Required to sync with Google Merchant Center and Google Ads.',
+						'google-listings-and-ads'
+					) }
+					<div>
```
@ankitrox No need to wrap the `CheckboxControl` within a `div`. `CheckboxControl` already outputs a `div` rendering the one you added unecessary.

asvinb commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,116 @@
+						<CheckboxControl
+							id={ checkboxId }
+							checked={ termsAccepted }
+							onChange={ ( val ) => setTermsAccepted( !! val ) }
```
```suggestion
							onChange={ setTermsAccepted }
```

asvinb commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,116 @@
+							id={ checkboxId }
+							checked={ termsAccepted }
+							onChange={ ( val ) => setTermsAccepted( !! val ) }
+							disabled={ disabled }
```
@ankitrox Can you let me know when this is expected?

ankitrox replied:
We should not be able to interact with checkbox or any other combo card component till jetpack is connected.

asvinb replied:
I see. Sounds good. I thought having `disabled` on `AccountCard` would be enough to prevent any pointer events, thus preventing any clicks within the disabled card. But that's fine for now.

asvinb commented on js/src/components/google-combo-account-card/index.scss:
```diff
@@ -0,0 +1,10 @@
+		}
+	}
+	p {
+		color: #757575;
```
```suggestion
		color: $gray-700;
```

asvinb replied:
This variable is available in the following file:
https://github.com/woocommerce/google-listings-and-ads/blob/develop/js/src/css/abstracts/_colors.scss#L2

which equals to #757575 https://github.com/WordPress/gutenberg/blob/33e2fadeebe32a64a946106488120b6f559d3deb/packages/base-styles/_colors.scss#L11

asvinb commented on js/src/components/google-combo-account-card/index.scss:
```diff
@@ -0,0 +1,10 @@
+			display: inline-block;
+		}
+	}
+	p {
```
```suggestion

	p {
```
Leave an empty line between the selectors please.

asvinb commented on js/src/components/google-combo-account-card/index.scss:
```diff
@@ -0,0 +1,10 @@
+.google-combo-account-card {
+	.gla-account-card__description {
```
@ankitrox I think this won't be necessary if we use the `label` prop of `CheckboxControl`

ankitrox replied:
We can't use `label` prop for the reason specified above.

asvinb reviewed (changes requested):
@ankitrox I left a few comments. Can you kindly check them out and let me know what you think please?
Thanks!

ankitguptaindia commented:
**QA Test Report**
----

I have completed the QA for this PR and confirmed that the requirements align with the details provided in the connected issue (FR ticket). 
Below are the test cases I performed, all of which are functioning as expected:

1. **Google Accounts Card Disabled if Store Not Connected to WP.com** - If the store is not connected to WP.com, the Google account card should be disabled. ✅

   ![Screenshot 2024-09-13 at 5 05 07 PM](https://github.com/user-attachments/assets/285c9ff4-0c52-42af-8f4a-434b80bd8670)

---

2. **T&C Acceptance Required**  - The merchant should not be able to connect their account until the "Accept T&C" checkbox is checked.   ✅

   ![Screenshot 2024-09-13 at 5 21 20 PM](https://github.com/user-attachments/assets/78bfbbcf-6002-42b1-8a84-0766fdc12bd0)

----

3. **External Links Open in New Tabs** - [Merchant Center](https://support.google.com/merchants/answer/160173) and [Google Ads](https://support.google.com/adspolicy/answer/54818) links should be open in new tabs.  ✅


-----

4. **Button Behavior After T&C Acceptance** - After the merchant accepts the T&C checkbox, the 'Continue' button should be disabled, and only the 'Connect' button should be enabled.   ✅




   ![Screenshot 2024-09-13 at 5 37 04 PM](https://github.com/user-attachments/assets/40d32813-05d0-4a76-9a36-05b7ab5ac6a9)

----

5. **Access Screen for Existing Google Account**  - If the merchant uses an existing Google account that already has the required access, the access request screen shpuld not appear.  ✅

----

6. **Access Request Message**  - If the merchant doesn't provide full access or only grants partial access, the card shows a message requesting full access to proceed. ✅

   ![Screenshot 2024-09-13 at 5 18 34 PM](https://github.com/user-attachments/assets/5b3a364d-55e0-4062-bbcb-33946a44df66)

---

7. **Access Screen for New Google Account**   If the merchant uses a new Google account, the access request screen should be appear.  ✅

---


**Quick Demo:** 

https://github.com/user-attachments/assets/b5ae27f3-8a04-4886-ba22-1988d4eb9300

ankitguptaindia commented:
@ankitrox Can we align 'Ads' in the same line with 'Google'? 

![image](https://github.com/user-attachments/assets/949df080-00aa-426c-b46c-fd098b7af0a8)


--

QA Note- Will test this PR again quickly after completion of CR.

ankitrox commented:
> @ankitrox Can we align 'Ads' in the same line with 'Google'?
> 
> ![image](https://private-user-images.githubusercontent.com/6384299/367317908-949df080-00aa-426c-b46c-fd098b7af0a8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYyMzc4MDksIm5iZiI6MTcyNjIzNzUwOSwicGF0aCI6Ii82Mzg0Mjk5LzM2NzMxNzkwOC05NDlkZjA4MC0wMGFhLTQyNmMtYjQ2Yy1mZDA5OGI3YWYwYTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTNUMTQyNTA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Zjg4YTNiNjViZGQ0YzU1MWQyNjI4NjI2MzRjODcxM2I1YTFkNzE5NzJhNjVmZDg1MWNjMjYyODBiYWJjZTc0MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8JEubvdZobuuPQihDHoFo_3AqtY12HQqtWIKjVs1nqQ)
> 
> --
> 
> QA Note- Will test this PR again quickly after completion of CR.

@ankitguptaindia This has been fixed with the new `label` prop for `CheckboxControl` component.

ankitrox committed:
Use label prop for checkbox.

ankitrox committed:
Styling fixes.

asvinb commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,111 @@
+		useGoogleConnectFlow( pageName );
+	const [ termsAccepted, setTermsAccepted ] = useState( false );
+
+	const checkboxId = 'gla-account-card-terms-conditions';
```
@ankitrox Sorry I just noticed that. `checkboxId` is no longer used. Can you remove it please?

asvinb reviewed (approved):
@ankitrox Changes LGTM. Just a one thing to fix here:
https://github.com/woocommerce/google-listings-and-ads/pull/2601/files#r1759081955
and you can send to QA. 
Thanks!

ankitrox committed:
Remove unused variable.

ankitrox committed:
Update connect-google-combo-account-card.js

ankitguptaindia commented:
**QA/Test Report-** ✅ 
------------

**Testing Environment -** 

<Details>

* WordPress: 6.6.2
* Theme active on store: Twenty Twenty-Four Version: 1.2
* WooCommerce - Version 9.3.1
* PHP: 8.3
* Web Server: Nginx
* Browser: Chrome - Version 128
* OS: macOS Sonoma 14.6.1 

</Details>



**Test Results -**  Verified all cases after completion of CR. PR changes working as described in AC. 

Detailed Test cases status: https://github.com/woocommerce/google-listings-and-ads/pull/2601#issuecomment-2348890281



**Functional Demo /  Screencast -** 

**Google Card workflow when has all permission:** 


https://github.com/user-attachments/assets/76947812-c0eb-49f8-a456-81fbcd61f938


**Google Card workflow when has limited permission:** 

https://github.com/user-attachments/assets/dbe90d53-9686-46f6-99a0-f83b4cb9eaf7


_This PR is ready for the next step, we're discussing some minor concerns related to the button color status over [Slack](https://fueled10up.slack.com/archives/C060DC0NXNG/p1726486649650889), Will update/create this/new PR if the issue is related to current changes._

eason9487 commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,108 @@
+	return (
+		<AccountCard
+			appearance={ APPEARANCE.GOOGLE }
+			className="google-combo-account-card"
```
Usually the CSS class name would be closer to the component name it attaches to unless it is/will be a high-level abstraction CSS class to be attached to more similar components.

Additionally, it would be better to add a prefix like `gla-` or `g4w-`.

ankitrox replied:
Updated the class name to `gla-connect-google-combo-account-card`.

eason9487 commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,108 @@
+						onChange={ setTermsAccepted }
+						disabled={ disabled }
+					/>
+					<p className="google-combo-account-card-description">
```
The CSS class naming concern is the same as the above `className="google-combo-account-card"`.

Also, the *description* in the name might be better to be adjusted, since it doesn't actually apply to the entire description in the account card.

ankitrox replied:
Updated the class name to `gla-connect-google-combo-account-card-instruction`

eason9487 commented on js/src/components/google-combo-account-card/index.scss:
Using `connect-google-combo-account-card.scss` as its file name would be better if this is only for the `ConnectGoogleComboAccountCard` component.

eason9487 commented on js/src/components/google-combo-account-card/index.js:
```diff
@@ -0,0 +1,34 @@
+import RequestFullAccessGoogleAccountCard from '../google-account-card/request-full-access-google-account-card';
+import { ConnectedGoogleAccountCard } from '../google-account-card';
+import ConnectGoogleComboAccountCard from './connect-google-combo-account-card';
+import './index.scss';
```
Using the `connect-google-combo-account-card.js` to import this style would be better if this is only for the `ConnectGoogleComboAccountCard` component.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -172,22 +172,39 @@ test.describe( 'Set up accounts', () => {
+
+			await setUpAccountsPage.goto();
+
+			const connectButton = page
+				.getByRole( 'button', {
+					name: 'Connect',
+					exact: true,
+				} )
+				.nth( 1 );
```
Suggest using a more specific way to locate the button.

```suggestion
			const connectButton = setUpAccountsPage
				.getGoogleAccountCard()
				.getByRole( 'button', { name: 'Connect' } );
```

ankitrox replied:
Thanks for the suggestion, it seems more appropriate.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -196,6 +213,8 @@ test.describe( 'Set up accounts', () => {
 				baseURL + 'google_auth'
 			);
 
+			setUpAccountsPage.getTermsCheckbox().check();
```
Awaiting the async action would be better.

```suggestion
			await setUpAccountsPage.getTermsCheckbox().check();
```

eason9487 commented on tests/e2e/utils/mock-requests.js:
```diff
@@ -392,6 +392,17 @@ export default class MockRequests {
 	}
 
+	/**
+	 * Mock Jetpack as connected.
```
The JSDoc description is incorrect.

ankitrox replied:
Thank you. Corrected it.

eason9487 commented on tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js:
```diff
@@ -367,4 +367,13 @@ export default class SetUpAccountsPage extends MockRequests {
+	 * @return {import('@playwright/test').Locator} Terms checkbox.
+	 */
+	getTermsCheckbox() {
+		return this.page.locator( '#gla-account-card-terms-conditions' );
```
This locator couldn't locate the Terms checkbox.

eason9487 commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.js:
```diff
@@ -0,0 +1,108 @@
+					/>
+					<p className="google-combo-account-card-description">
+						<em>
+							{ createInterpolateElement(
+								__(
+									'You will be prompted to give WooCommerce access to your Google account. Please check all the checkboxes to give WooCommerce all required permissions. <link>Read more</link>',
+									'google-listings-and-ads'
+								),
+								{
+									link: readMoreLink,
+								}
+							) }
```
Maybe consider using the `helper` prop of `AccountCard` component to render this content without wrapping it by `<p>` and `<em>`. In this way, it will only need to adjust `font-size` style via the `.gla-account-card__helper` CSS class without adding the `google-combo-account-card-description` CSS class.

For example:

```scss
.gla-connect-google-combo-account-card {
	.gla-account-card__helper {
		font-size: $gla-font-base;
	}
}
```

ankitrox replied:
Thank you @eason9487 

I have added the text as helper prop.

eason9487 reviewed (changes requested):
Thanks for the work. Left some comments that may need to be addressed.

Furthermore:
- The e2e testing couldn't pass - https://github.com/woocommerce/google-listings-and-ads/actions/runs/10897655273/job/30239450725
- Now these two assertions are false negatives. It would be nicer to be removed together.
https://github.com/woocommerce/google-listings-and-ads/blob/8a152ba5a72427fb239d30fdc4fb35908c13fa40/tests/e2e/specs/setup-mc/step-1-accounts.test.js#L69-L73

ankitrox committed:
Update css class names and fix tests.

ankitrox committed:
Merge branch 'feature/2566-google-combo-card' of https://github.com/woocommerce/google-listings-and-ads into feature/2566-google-combo-card

ankitrox committed:
Fix: e2e checkbox locator.

ankitrox commented:
> Thanks for the work. Left some comments that may need to be addressed.
> 
> Furthermore:
> 
> * The e2e testing couldn't pass - https://github.com/woocommerce/google-listings-and-ads/actions/runs/10897655273/job/30239450725
> * Now these two assertions are false negatives. It would be nicer to be removed together.
>   https://github.com/woocommerce/google-listings-and-ads/blob/8a152ba5a72427fb239d30fdc4fb35908c13fa40/tests/e2e/specs/setup-mc/step-1-accounts.test.js#L69-L73

Thank you @eason9487 for reviewing the PR.

I've removed the redundant checks in e2e and also fixed the failing tests.

ankitrox committed:
Fix: JS lint.

asvinb commented:
Changes LGTM @ankitrox .
Assigning to @eason9487 to take another look.

eason9487 commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.scss:
```diff
@@ -0,0 +1,11 @@
+.gla-connect-google-combo-account-card {
+
+	.gla-account-card__helper {
+		color: $gray-700;
```
This style can be omitted as the same color has been set in `.gla-account-card__helper`.

eason9487 commented on js/src/components/google-combo-account-card/connect-google-combo-account-card.scss:
```diff
@@ -0,0 +1,11 @@
+		color: $gray-700;
+		font-size: $gla-font-base;
+
+		p {
+			margin-bottom: 0;
+		}
```
This style should not be needed because there is no `<p>` within it.

eason9487 reviewed (approved):
Thanks for the updates! LGTM.

There are two unnecessary styles in `connect-google-combo-account-card.scss` to be removed, and I believe the removal won't need another round of code reviews from Woo. I will be approving this RP in advance.

ankitrox committed:
Remove redundant css rules.
